File: jarvis/backend/app/core/telemetry/monitor.py
"""System telemetry sampling via ``psutil``.

``TelemetryMonitor.sample()`` returns a :class:`~app.models.schemas.Telemetry`
snapshot. Network throughput (``net_up`` / ``net_down``) is reported as an
instantaneous rate in **KB/s**, computed from the delta between this sample's
cumulative byte counters and the previous sample's -- never as a raw
cumulative total, since a growing total is not useful telemetry to stream to
a HUD every second.

Every individual sensor read is wrapped so a platform missing that sensor
(no battery, no exposed CPU temperature, a sandboxed container with a
restricted ``/proc``) degrades that one field to ``None``/``0`` instead of
raising out of ``sample()``. ``psutil`` itself is a hard dependency of this
module (it is listed in ``requirements.txt``), but every *individual call*
into it is still defensively wrapped, since ``psutil``'s per-platform sensor
coverage is inconsistent (e.g. ``sensors_temperatures`` is Linux-only and
``sensors_battery`` is desktop-only).
"""

from __future__ import annotations

import asyncio
import logging
import time
from dataclasses import dataclass
from typing import Awaitable, Callable

from app.models.schemas import Telemetry

logger = logging.getLogger(__name__)

_BYTES_PER_KB = 1024.0


@dataclass
class _NetSnapshot:
    bytes_sent: int
    bytes_recv: int
    timestamp: float


class TelemetryMonitor:
    """Stateful sampler: tracks the previous network counters to derive rates.

    One instance is expected to live for the lifetime of a connection /
    process (see ``run``), since the very first ``sample()`` has no prior
    reading to diff against and reports ``net_up``/``net_down`` as ``0.0``.
    """

    def __init__(self) -> None:
        self._prev_net: _NetSnapshot | None = None

    def sample(self) -> Telemetry:
        """Take one point-in-time telemetry reading. Never raises."""
        cpu = self._read_cpu_percent()
        mem = self._read_mem_percent()
        disk = self._read_disk_percent()
        net_up, net_down = self._read_net_rates()
        battery = self._read_battery_percent()
        uptime_s = self._read_uptime_s()
        processes = self._read_process_count()
        temp_c = self._read_cpu_temp()

        return Telemetry(
            cpu=cpu,
            mem=mem,
            disk=disk,
            net_up=net_up,
            net_down=net_down,
            battery=battery,
            uptime_s=uptime_s,
            processes=processes,
            temp_c=temp_c,
        )

    async def run(
        self,
        on_sample: Callable[[Telemetry], Awaitable[None]],
        interval_ms: int = 1000,
    ) -> None:
        """Broadcast loop: sample on a fixed interval and await ``on_sample``.

        Runs until cancelled (the normal way callers stop it -- e.g. a
        WebSocket disconnect cancelling the task). A failure in ``on_sample``
        (e.g. a closed socket) is logged and the loop continues rather than
        dying silently; a failure in ``sample()`` itself cannot happen by
        contract, but is guarded anyway for defense in depth.
        """
        interval_s = max(0.05, interval_ms / 1000.0)
        while True:
            try:
                telemetry = self.sample()
                await on_sample(telemetry)
            except asyncio.CancelledError:
                raise
            except Exception:  # pragma: no cover - defensive, caller-supplied callback
                logger.warning("Telemetry broadcast callback raised; continuing loop", exc_info=True)
            await asyncio.sleep(interval_s)

    # -- individual sensor reads, each independently fail-safe -------------

    @staticmethod
    def _read_cpu_percent() -> float:
        try:
            import psutil

            # interval=None -> non-blocking, compares against the last call
            # (0.0 on this process's very first call, which is fine: 0..100).
            return float(psutil.cpu_percent(interval=None))
        except Exception:
            logger.warning("Failed to read CPU percent", exc_info=True)
            return 0.0

    @staticmethod
    def _read_mem_percent() -> float:
        try:
            import psutil

            return float(psutil.virtual_memory().percent)
        except Exception:
            logger.warning("Failed to read memory percent", exc_info=True)
            return 0.0

    @staticmethod
    def _read_disk_percent() -> float:
        try:
            import psutil

            return float(psutil.disk_usage("/").percent)
        except Exception:
            logger.warning("Failed to read disk usage", exc_info=True)
            return 0.0

    def _read_net_rates(self) -> tuple[float, float]:
        """Return ``(upload_kb_s, download_kb_s)`` computed from counter deltas."""
        try:
            import psutil

            counters = psutil.net_io_counters()
            now = time.monotonic()
            current = _NetSnapshot(
                bytes_sent=counters.bytes_sent, bytes_recv=counters.bytes_recv, timestamp=now
            )

            previous = self._prev_net
            self._prev_net = current

            if previous is None:
                return 0.0, 0.0

            elapsed = current.timestamp - previous.timestamp
            if elapsed <= 0:
                return 0.0, 0.0

            sent_delta = max(0, current.bytes_sent - previous.bytes_sent)
            recv_delta = max(0, current.bytes_recv - previous.bytes_recv)

            net_up = (sent_delta / _BYTES_PER_KB) / elapsed
            net_down = (recv_delta / _BYTES_PER_KB) / elapsed
            return round(net_up, 2), round(net_down, 2)
        except Exception:
            logger.warning("Failed to read network counters", exc_info=True)
            return 0.0, 0.0

    @staticmethod
    def _read_battery_percent() -> float | None:
        try:
            import psutil

            battery = psutil.sensors_battery()
            if battery is None:
                return None
            return float(battery.percent)
        except Exception:
            # AttributeError on platforms without sensors_battery at all.
            logger.debug("Battery sensor unavailable on this platform", exc_info=True)
            return None

    @staticmethod
    def _read_uptime_s() -> int:
        try:
            import psutil

            return max(0, int(time.time() - psutil.boot_time()))
        except Exception:
            logger.warning("Failed to read system uptime", exc_info=True)
            return 0

    @staticmethod
    def _read_process_count() -> int:
        try:
            import psutil

            return len(psutil.pids())
        except Exception:
            logger.warning("Failed to read process count", exc_info=True)
            return 0

    @staticmethod
    def _read_cpu_temp() -> float | None:
        try:
            import psutil

            temps = psutil.sensors_temperatures()
            if not temps:
                return None
            for readings in temps.values():
                if readings:
                    return float(readings[0].current)
            return None
        except Exception:
            # AttributeError on platforms without sensors_temperatures at all
            # (notably macOS), or PermissionError in some sandboxes.
            logger.debug("CPU temperature sensor unavailable on this platform", exc_info=True)
            return None

File: jarvis/server/src/telemetry/monitor.ts
/**
 * System telemetry sampling via `systeminformation`.
 *
 * `TelemetryMonitor.sample()` returns one `Telemetry` snapshot (see
 * `types.ts`). `net_up`/`net_down` are instantaneous **KB/s rates**, derived
 * from the delta between this sample's cumulative byte counters and the
 * previous sample's — never a raw cumulative total, since a monotonically
 * growing total is not useful telemetry to stream to a HUD every second.
 *
 * Every individual sensor read is independently guarded: a machine with no
 * battery or no exposed CPU temperature degrades that one field to `null`
 * (or `0` for counters) and keeps every other field live. `sample()` itself
 * never rejects/throws.
 *
 * Ported from `backend/app/core/telemetry/monitor.py`.
 */

import si from 'systeminformation';
import type { Telemetry } from '../types.js';

const BYTES_PER_KB = 1024;

interface NetSnapshot {
  bytesSent: number;
  bytesRecv: number;
  timestampMs: number;
}

export class TelemetryMonitor {
  private prevNet: NetSnapshot | null = null;

  /** Take one point-in-time telemetry reading. Never throws. */
  async sample(): Promise<Telemetry> {
    const [cpu, mem, disk, net, battery, uptimeS, processes, temp] = await Promise.all([
      this.readCpuPercent(),
      this.readMemPercent(),
      this.readDiskPercent(),
      this.readNetRates(),
      this.readBatteryPercent(),
      this.readUptimeS(),
      this.readProcessCount(),
      this.readCpuTemp(),
    ]);

    return {
      cpu,
      mem,
      disk,
      net_up: net.up,
      net_down: net.down,
      battery,
      uptime_s: uptimeS,
      processes,
      temp_c: temp,
    };
  }

  /**
   * Sample on a fixed interval and invoke `onSample`. Runs until `signal` is
   * aborted. A failure in `onSample` (e.g. a closed socket) is swallowed and
   * the loop continues rather than dying silently.
   */
  async run(
    onSample: (telemetry: Telemetry) => void | Promise<void>,
    intervalMs: number,
    signal?: AbortSignal,
  ): Promise<void> {
    const interval = Math.max(50, intervalMs);
    while (!signal?.aborted) {
      try {
        const telemetry = await this.sample();
        await onSample(telemetry);
      } catch (err) {
        console.warn('[telemetry] broadcast callback failed, continuing loop', err);
      }
      await sleep(interval, signal);
    }
  }

  // -- individual sensor reads, each independently fail-safe ---------------

  private async readCpuPercent(): Promise<number> {
    try {
      const load = await si.currentLoad();
      return round2(load.currentLoad);
    } catch (err) {
      console.warn('[telemetry] failed to read CPU percent', err);
      return 0;
    }
  }

  private async readMemPercent(): Promise<number> {
    try {
      const mem = await si.mem();
      if (mem.total <= 0) return 0;
      return round2(((mem.total - mem.available) / mem.total) * 100);
    } catch (err) {
      console.warn('[telemetry] failed to read memory percent', err);
      return 0;
    }
  }

  private async readDiskPercent(): Promise<number> {
    try {
      const layout = await si.fsSize();
      if (!layout.length) return 0;
      // Weighted average across mounted filesystems, mirroring a single
      // root-usage percentage without assuming a "/" mount exists (Windows).
      const totalSize = layout.reduce((sum, d) => sum + (d.size || 0), 0);
      const totalUsed = layout.reduce((sum, d) => sum + (d.used || 0), 0);
      if (totalSize <= 0) return 0;
      return round2((totalUsed / totalSize) * 100);
    } catch (err) {
      console.warn('[telemetry] failed to read disk usage', err);
      return 0;
    }
  }

  private async readNetRates(): Promise<{ up: number; down: number }> {
    try {
      const stats = await si.networkStats();
      const bytesSent = stats.reduce((sum, s) => sum + (s.tx_bytes || 0), 0);
      const bytesRecv = stats.reduce((sum, s) => sum + (s.rx_bytes || 0), 0);
      const now = Date.now();
      const current: NetSnapshot = { bytesSent, bytesRecv, timestampMs: now };

      const previous = this.prevNet;
      this.prevNet = current;

      if (previous === null) return { up: 0, down: 0 };

      const elapsedS = (current.timestampMs - previous.timestampMs) / 1000;
      if (elapsedS <= 0) return { up: 0, down: 0 };

      const sentDelta = Math.max(0, current.bytesSent - previous.bytesSent);
      const recvDelta = Math.max(0, current.bytesRecv - previous.bytesRecv);

      return {
        up: round2(sentDelta / BYTES_PER_KB / elapsedS),
        down: round2(recvDelta / BYTES_PER_KB / elapsedS),
      };
    } catch (err) {
      console.warn('[telemetry] failed to read network counters', err);
      return { up: 0, down: 0 };
    }
  }

  private async readBatteryPercent(): Promise<number | null> {
    try {
      const battery = await si.battery();
      if (!battery.hasBattery) return null;
      return round2(battery.percent);
    } catch (err) {
      console.debug('[telemetry] battery sensor unavailable on this platform', err);
      return null;
    }
  }

  private async readUptimeS(): Promise<number> {
    try {
      return Math.max(0, Math.floor(si.time().uptime));
    } catch (err) {
      console.warn('[telemetry] failed to read system uptime', err);
      return 0;
    }
  }

  private async readProcessCount(): Promise<number> {
    try {
      const procs = await si.processes();
      return procs.all ?? 0;
    } catch (err) {
      console.warn('[telemetry] failed to read process count', err);
      return 0;
    }
  }

  private async readCpuTemp(): Promise<number | null> {
    try {
      const temp = await si.cpuTemperature();
      if (temp.main === null || temp.main === undefined || Number.isNaN(temp.main) || temp.main < 0) {
        return null;
      }
      return round2(temp.main);
    } catch (err) {
      console.debug('[telemetry] CPU temperature sensor unavailable on this platform', err);
      return null;
    }
  }
}

function round2(n: number): number {
  return Math.round(n * 100) / 100;
}

function sleep(ms: number, signal?: AbortSignal): Promise<void> {
  return new Promise((resolve) => {
    const timer = setTimeout(resolve, ms);
    signal?.addEventListener(
      'abort',
      () => {
        clearTimeout(timer);
        resolve();
      },
      { once: true },
    );
  });
}
